experiments-core/include/foedus/ycsb/ycsb.hpp

```diff
@@ -0,0 +1,322 @@
+ *
+ * Masstree is used as the default storage for all benchmarks; a hashtable variant is
+ * available for A -- D only (E has to use Masstree for scans).
+ *
+ * Two major differences from the Java reference implementation:
+ *
+ * 1. Key format
+ * Yahoo!: "user" + hash(integer counter)
+ * FOEDUS: "user" + hash(worker ID | integer counter)
+ * The counter in the lower bits is thread-local. For now inserters only insert to their own key
+ * space, separated by the worker ID. Later we might modify this to allow inserting to others'
+ * key space. Threads doing reads and updates randomly choose a worker ID as the key's high bits,
+ * then they randomly choose a worker ID and take a look at (no synchronization) its local key
+ * counter value (k), randomly choose an integer within [0, k] as the low bits.
+ *
+ * 2. RNGs
+ * So far we have only uniform RNGs. TODO(tzwang): add zipfian and others.
+ *
+ * [Cooper10] Benchmarking cloud serving systems with YCSB, SoCC '10.
+ */
+namespace foedus {
+namespace ycsb {
+const uint32_t kMaxUnexpectedErrors = 1;
+
+/* Number of bytes in each field */
+const uint32_t kFieldLength = 100;
+
+/* Number of fields per record */
+const uint32_t kFields = 10;
+
+/* Record size */
+const uint64_t kRecordSize = kFieldLength * kFields;
+
+// TODO(tzwang): make it a cmd argument
+const uint64_t kInitialUserTableSize= 10000;
+
+const uint32_t kMaxWorkers = 1024;
+
+const int kKeyPrefixLength = 4;  // "user" without \0
+const assorted::FixedString<kKeyPrefixLength> kKeyPrefix("user");
+
+const int32_t kKeyMaxLength = kKeyPrefixLength + 32;  // 4 bytes "user" + 32 chars for numbers
+class YcsbKey {
+ private:
+  assorted::FixedString<kKeyMaxLength> data_;
+
+ public:
+  YcsbKey() {}
+  YcsbKey& next(uint32_t worker_id, uint32_t* local_counter);
+  YcsbKey& build(uint32_t high_bits, uint32_t low_bits);
+
+  const char *ptr() const {
+    return data_.data();
+  }
+
+  uint32_t size() const {
+    return data_.length();
+  }
+};
+
+struct YcsbRecord {
+  char data_[kFieldLength][kFields];
+  explicit YcsbRecord(char value);
+  YcsbRecord() {}
+  static void initialize_field(char *field);
+};
+
+/**
+ * Channel between the driver process/thread and clients process/thread.
+ * If the driver spawns client processes, this is allocated in shared memory.
+ * This ``channel'' controls/synchoronizes worker threads.
+ */
+class YcsbClientTask;
+struct YcsbClientChannel {
+  void initialize() {
+    start_rendezvous_.initialize();
+    exit_nodes_.store(0);
+    stop_flag_.store(false);
+    nr_workers_ = 0;
+    workers_mutex_.initialize();
+  }
+  void uninitialize() {
+    start_rendezvous_.uninitialize();
+    workers_mutex_.uninitialize();
+  }
+  uint32_t peek_local_key_counter(uint32_t worker_id);
+
+  soc::SharedRendezvous start_rendezvous_;
+  std::atomic<uint16_t> exit_nodes_;
+  std::atomic<bool> stop_flag_;
+  uint32_t nr_workers_;
+  soc::SharedMutex workers_mutex_;
+  std::vector<YcsbClientTask> workers;
```

Comment:
std::vector here? Does it run on fork_workers=true? std::vector allocates memory in heap, so it will break when they are running in different processes. Also, workers_